odd-platform-api/src/main/java/com/provectus/oddplatform/service/metric/extractors/MetricExtractorImpl.java

```diff
@@ -0,0 +1,146 @@
+            .flatMap(de -> de.getDataSet().getFieldList().stream()
+                .filter(f -> f.getStats() != null)
+                .flatMap(f -> extractFieldMetrics(de.getOddrn(), f)));
+
+        return gaugeStream(metricStream);
+    }
+
+    private Stream<MetricData> gaugeStream(final Stream<Pair<MetricDataTriplet, ? extends PointData>> metricStream) {
+        return metricStream
+            .filter(p -> p.getRight() != null)
+            // TODO: get rid of intermediate collector
+            .collect(Collectors.groupingBy(Pair::getLeft))
+            .entrySet()
+            .stream()
+            .map(e -> Pair.of(e.getKey(), e.getValue().stream().map(Pair::getRight).collect(Collectors.toList())))
+            .map(p -> gauge(p.getLeft(), p.getRight()));
+    }
+
+    private Stream<Pair<MetricDataTriplet, LongPointData>> buildMetrics(final IngestionTaskRun taskRun) {
+        final long duration = Duration.between(taskRun.getEndTime(), taskRun.getStartTime()).getSeconds();
+
+        final long status = taskRun.getStatus().equals(IngestionTaskRunStatus.SUCCESS) ? 1L : 0L;
+
+        final Attributes attributes =
+            Attributes.of(AttributeKey.stringKey("job_oddrn"), taskRun.getDataEntityOddrn());
+
+        return Stream.of(
+            Pair.of(MetricDataTriplet.TASK_RUN_DURATION, longPointData(duration, attributes)),
+            Pair.of(MetricDataTriplet.DF_NULLS_COUNT, longPointData(status, attributes))
+        );
+    }
+
+    private Pair<MetricDataTriplet, LongPointData> buildMetrics(final DataEntityIngestionDto entity) {
+        final Attributes attributes = Attributes.of(AttributeKey.stringKey("entity_oddrn"), entity.getOddrn());
+
+        return Pair.of(
+            MetricDataTriplet.ROWS_COUNT,
+            longPointData(entity.getDataSet().getRowsCount(), attributes)
+        );
+    }
+
+    private Stream<Pair<MetricDataTriplet, ? extends PointData>> extractFieldMetrics(final String entityOddrn,
+                                                                                     final DataSetField field) {
+        return Stream.of(
+            binaryExtractor.extract(entityOddrn, field.getOddrn(), field.getStats().getBinaryStats()),
+            booleanExtractor.extract(entityOddrn, field.getOddrn(), field.getStats().getBooleanStats()),
+            complexExtractor.extract(entityOddrn, field.getOddrn(), field.getStats().getComplexStats()),
+            numberExtractor.extract(entityOddrn, field.getOddrn(), field.getStats().getNumberStats()),
+            stringExtractor.extract(entityOddrn, field.getOddrn(), field.getStats().getStringStats()),
+            integerExtractor.extract(entityOddrn, field.getOddrn(), field.getStats().getIntegerStats()),
+            dateTimeExtractor.extract(entityOddrn, field.getOddrn(), field.getStats().getDatetimeStats())
+        ).flatMap(identity());
+    }
+
+    @SuppressWarnings("unchecked")
```

Comment:
unavoidable?

Reply:
At the moment. It's possible that will be rewritten in the nearest feature